source/parser.hera

```diff
@@ -3719,32 +3680,93 @@ JSXAttributeInitializer
 # https://facebook.github.io/jsx/#prod-JSXAttributeValue
 JSXAttributeValue
   # https://facebook.github.io/jsx/#prod-JSXDoubleStringCharacters
-  /"[^"]*"/
   # https://facebook.github.io/jsx/#prod-JSXSingleStringCharacters
-  /'[^']*'/
+  # But double-quoted strings may turn into interpolation,
+  # and some triple-quoted strings may not use interpolation.
+  # So try to match but skip here if we find interpolation,
+  # and add parens in InlineJSXAttributeValue otherwise.
+  StringLiteral ->
+    if (module.isTemplateLiteral($1)) return $skip
+    return $1
   # NOTE: Using ExtendedExpression to allow If/Switch expressions
   OpenBrace ExtendedExpression __ CloseBrace
   JSXElement
   JSXFragment
   InsertInlineOpenBrace InlineJSXAttributeValue InsertCloseBrace
 
 # JSX shorthand to avoid explicit braces when unnecessary.
-# Subset of PrimaryExpression; omissions documented in comments below.
 InlineJSXAttributeValue
+  InlineJSXUnaryExpression InlineJSXBinaryOpRHS* ->
+    if ($2.length)
+      return module.processBinaryOpExpression($0)
+    return $1
+
+# BinaryOpRHS without whitespace and without ExpressionizedStatement
+InlineJSXBinaryOpRHS
+  Nothing BinaryOp Nothing ( ParenthesizedAssignment / InlineJSXUnaryExpression )
+
+# JSXUnaryExpression, with InlineJSX prefixes and no Do (which has whitespace)
+InlineJSXUnaryExpression
+  InlineJSXUnaryOp*:pre InlineJSXUpdateExpression:exp InlineJSXUnaryPostfix?:post ->
+    return module.processUnaryExpression(pre, exp, post)
+
+# UnaryOp, restricted to whitespace-free symbol operators
+InlineJSXUnaryOp
+  /[!~+-](?!\s|[!~+-]*&)/ ->
+    return { $loc, token: $0 }
+
+# UnaryPostfix, restricted to whitespace-free symbol operators
+InlineJSXUnaryPostfix
+  QuestionMark
+
+# UpdateExpression, with LeftHandSideExpression -> InlineJSXCallExpression
+InlineJSXUpdateExpression
+  UpdateExpressionSymbol UnaryExpression
+  InlineJSXCallExpression UpdateExpressionSymbol? ->
+    if ($2) return $0
+    return $1
+
+# CallExpression, with only explicit function calls.
+# This also acts as a replacement for LeftHandSideExpression
+# (we don't allow New because that has whitespace).
+InlineJSXCallExpression
+  "super" ExplicitArguments
+  "import" OpenParen ExtendedExpression __ CloseParen
+  InlineJSXMemberExpression InlineJSXCallExpressionRest* ->
+    if ($2.length) return $0
+    return $1
+
+# CallExpressionRest, with only explicit function calls.
+InlineJSXCallExpressionRest
+  MemberExpressionRest
+  TemplateLiteral
+  ( OptionalShorthand / NonNullAssertion )? ExplicitArguments
+
+# MemberExpression, with PrimaryExpression -> InlineJSXPrimaryExpression
+InlineJSXMemberExpression
+  InlineJSXPrimaryExpression MemberExpressionRest* ->
```

Comment:
There is an indented construction inside `MemberExpressionRest` to handle things like:

```
f a
  .b
f a
.b
---
f(a
.b)
f(a)
.b
```

We probably don't want to allow nested indentation accese here.